Update simulated geth client wrapper & LogPoller tests

reductionista opened the pull request:
In this PR:

1. Update `SimulatedBackendClient` to wrap simulated.Backend & simulated.Client ( instead of deprecated backends.SimulatedBackendClient )
2. Add `SetActiveClient()` to `SimulatedBackendClient` to simulate switching from one rpc server to another while running.
3. Add `RegisterHeadByNumberCallback` to `SimulatedBackendCilent`, to register a custom callback function to be called each time`HeadByNumber()` is called.
4. Add to `SimulateBackendClient` ability to run in "optimism mode", which behaves slightly different from geth: returns success rather than "block not found" if `FilterLogs()` is called with an invalid or unknown block hash. 
5. Update BackupPoller test to use the previous 3 features to simulate failing over.to an out-of-sync optimism rpc server during a race condition between pulling a particular block and its corresponding logs. (This is exactly the case which motivated the creation of Backup LogPoller.) This avoids the need to use rawdb, which is no longer accessible in geth.

6. Replace `markBlockAsFinalized()` and `markBlockAsFinalizedByHash()` in LogPoller tests with `finalizeThroughBlock()`.
7. Refactor logic in LogPoller tests previously using `markBlockAsFinalized(ByHash)` to make use of `finalizedThroughBlock()` instead.
8. Update some tests for new block timestamp behavior in Simulated Geth: automatic time increment between simulated blocks is 1ns now instead of 1s, and Client.AdjustTime() now automatically commits the current block after adjusting its timestamp
9. Update rest of tests in LogPoller and other packages to use new data types associated with (1.), instead of deprecated types

reductionista committed:
Update SimulatedBackendClient to wrap simulated.Backend & simulated.Client

( instead of deprecated backends.SimulatedBackendClient )

reductionista committed:
Update LogPoller helper

reductionista committed:
Add support for switching rpc clients in simulated geth

reductionista committed:
Fix TestLogPoller_BackupPollAndSaveLogsSkippingLogsThatAreTooOld

This test relied on markBlockFinalized() which has been replaced
with finalizedBlocksThrough(). Just needed some slight adjustments
to keep testing the same thing.

reductionista committed:
Fix TestLogPoller_ReorgDeeperThanFinality

reductionista committed:
Fix Test_PollAndSavePersistsFinalityInBlocks

reductionista committed:
Fix Test_PollAndQueryFinalizedBlocks

reductionista committed:
Fix chainreader & config poller tests

reductionista committed:
update listener_v2_log_listener_test

reductionista committed:
Re-run make generate, fix fluxmonitorv2 & ocr2keeper tests

reductionista committed:
Update keeper integration tests

reductionista committed:
Re-run make generate

reductionista committed:
Update BackupLogPoller test

Add RegisterHeadNumberCallback() to SimulatedBackendClient, so we can
trigger an rpc failover event just after reading a particular block, but
before the logs get read for that block. This is the race condition that
can happen on optimism chain which BackupPoller was designed to address

reductionista committed:
Update TestLogPoller_PollAndSaveLogsDeepReorg

reductionista committed:
Not sure how this was working before

Presumably, the older verison of simulated.Backend would fill in fake timestamps
instead of real ones?

reductionista commented:
Taking this out of draft status, because the most complex part of it--getting Backup LogPoller test to work--is done and passing. Most of the tests that broke due to not being able to mark blocks finalized any more have also been updated and are passing.  Just have a few more to go through, as well as updating the types in some tests. Seems like a good time to start getting some eyes on it.

jmank88 commented on core/chains/evm/logpoller/helper_test.go:
```diff
@@ -8,20 +8,21 @@ import (
 	"testing"
 	"time"
 
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient/simulated"
```
Let's group these with the other geth lines

jmank88 commented on core/gethwrappers/functions/generation/generated-wrapper-dependency-versions-do-not-edit.txt:
Did you mean to update these functions files? Looks like stray `make generate` effects that can be discarded

reductionista replied:
They can probably be discarded, I just figured it's better to have the latest `make generate` checked in than something out of date.

One reason I re-ran `make generate` was because `GETH_VERSION` was still set to the old version in some of them:
```
-GETH_VERSION: 1.13.8
-forwarder: ../../../contracts/solc/v0.8.19/KeystoneForwarder/KeystoneForwarder.abi ../../../contracts/solc/v0.8.19/KeystoneForwarder/KeystoneForwarder.bin ed9164cfe4619dff824b11df46b66f4c6834b2ca072923f10d9ebc57ce508ed8
-keystone_capability_registry: ../../../contracts/solc/v0.8.19/CapabilityRegistry/CapabilityRegistry.abi ../../../contracts/solc/v0.8.19/CapabilityRegistry/CapabilityRegistry.bin d4e0661491c2adc7f0d7553287c938fb32664b9f07770f6d57ae6511ce9884cd
-ocr3_capability: ../../../contracts/solc/v0.8.19/OCR3Capability/OCR3Capability.abi ../../../contracts/solc/v0.8.19/OCR3Capability/OCR3Capability.bin 9dcbdf55bd5729ba266148da3f17733eb592c871c2108ccca546618628fd9ad2
diff --git a/core/gethwrappers/operatorforwarder/generation/generated-wrapper-dependency-versions-do-not-edit.txt b/core/gethwrappers/operatorforwarder/generation/generated-wrapper-dependency-versions-do-not-edit.txt
index 1569801b3f..a30823b737 100644
--- a/core/gethwrappers/operatorforwarder/generation/generated-wrapper-dependency-versions-do-not-edit.txt
+++ b/core/gethwrappers/operatorforwarder/generation/generated-wrapper-dependency-versions-do-not-edit.txt
@@ -1,4 +1,4 @@
-GETH_VERSION: 1.13.8
+GETH_VERSION: 1.13.14
```
Should the above changes get backed out?

In the case of KeystoneForwarder, it was not only using the old version of geth, but the abi itself has changed. I can't remember for sure whether that was the main motivation--it might have been causing some test failures but not sure.

Looks like the only difference in functions_allow_list, functions_coordinator, and llo-feeds were the compiled binary itself, so I assume those can be ignored.

jmank88 replied:
CI already enforces that the latest generate version is used. If you are able to edit generated files without failing a check, then you found something that is not tracked correctly. There is a gap that old generated files which are not being regenerated are not noticed because they don't produce a diff.

jmank88 replied:
see: #13138
 > As a follow-up to prevent this from happening: Our CI process does not see these as problematic, since they are committed and do not produce a diff. We could add a step before make generate that removes all the generated mockery files to ensure that one which do not get re-generated produce a diff that fails CI.

jmank88 commented on core/chains/evm/client/simulated_backend_client.go:
```diff
@@ -66,20 +66,35 @@ var (
+// where success rather than an error code is returned when a call to FilterLogs() fails to find the block hash
+// requested. This combined with a failover event can lead to the "eventual consistency" behavior that Backup LogPoller
+// and other solutions were designed to recover from.
+func (c *SimulatedBackendClient) SetBackend(backend evmtypes.Backend, optimismMode bool) {
+	c.optimismMode = optimismMode
```
Would it make sense to pass ChainType here, so the API stays a bit more forward compatible, instead of being so closely tied to optimism? Can even leave the internals as-is for now:
```suggestion
func (c *SimulatedBackendClient) SetBackend(backend evmtypes.Backend, chainType evm.ChainType) {
	c.optimismMode = chainType == evm.ChainTypeOptimism
```

reductionista committed:
Update TestLogPoller_PollAndSaveLogs

reductionista committed:
Update TestLogPoller_Blocktimestamps

The new Simulated Geth made two changes which affected this test
1. The automatic time interval added to each new block is now 1ns instead of 1s
2. AdjustTime() now automatically calls Commit() so it no longer needs to be called aftewards

reductionista committed:
Update TestLogPoller_BackupPollAndSaveLogsWithPollerNotWorking

reductionista committed:
Address PR review comments

- Consolidate go-ethereum imports
- Remove extra geth-wrapper changes

reductionista committed:
Update types in vrf tests

reductionista committed:
Update keepers, fluxmointor, transmitter,ocr test types

reductionista committed:
re-generate KeystoneForwarder

reductionista committed:
Replace optimismMode with chainFamily enum

reductionista committed:
Update more types GenesisAlloc, GenesisAccount, llo

reductionista committed:
Fix some more compilation errors (fluxmonitor2, vrf, ocr2, functions)

reductionista committed:
Fix lint errors, remove unused logpoller test definitions

jmank88 commented on 2:
scratch file?

reductionista committed:
.

reductionista commented:
Present status:

- All tests in LogPoller package passing
- All deprecated types have been updated (in all packages), Lint is now passing

- Seeing many errors in the Core CI tests still, but nearly all of them are one of 3 things, which seem very related and possibly have the same root cause:
`no contract code at given address`
`Execution reverted`
`Execution reverted: Target is not a contract"

- One weird error in a core test, that happens in CI but passes locally:
--- FAIL: TestEVMForwarderPresenter_RenderTable (0.00s)
    forwarders_commands_test.go:48:
            Error Trace:    /home/runner/work/chainlink/chainlink/core/cmd/forwarders_commands_test.go:48
            Error:          "\n------------------------------------------------------\nID:         \nAddress:    0x4Fc0FeeD7E3F8d9bb6758d3084aa0ff6cFB5C486\nChain ID:   4\nCreated At: 2024-05-29T04:33:50Z\n------------------------------------------------------\n" does not contain "1"

- For some reason, this error causes the `Clean Go Tidy & Generate` CI check to fail:
cp: cannot create directory '../../operator_ui/..//core/web/assets': No such file or directory

I think the last one was what prompted me initially to run `make generate` locally, but I guess that didn't fix it. 

I'm not sure how many of these errors were already in the branch I'm trying to merge into, but seems like it might be about time to merge it and rebase that branch?

reductionista committed:
Run go-generate again on KeystoneForwarder

jmank88 commented on core/gethwrappers/keystone/generation/generated-wrapper-dependency-versions-do-not-edit.txt:
Can you restore this one?

jmank88 replied:
Er maybe not necessary for this case? I don't know these well.

reductionista replied:
Good catch! Not sure how that disappeared. For some reason if it's not there already, it just skips over it rather than trying to add it... so I restored the original (with GETH_VERSION 1.13.8 and then re-ran `make generate`, now it's updated it to GETH_VERSION 1.13.14

reductionista committed:
Re-run "make generate" one more time, this time without any failures aside from the // error at the end that's in CI as well.

reductionista committed:
Re-generate generation version db for keystone
